### ISSUE TITLE
How to use glDrawElements

### ISSUE THREAD

**Issue by FireBanana** [TheBoneJarmer/TilarGL#4]
glDrawElements doesn't seem to render anything, but glDrawArrays works fine in the context (Texture is rendered correctly on glDrawArrays as well).

Code in question here:

![image](https://user-images.githubusercontent.com/6471716/100161318-06ae5e80-2ed3-11eb-9286-223f56219ae8.png)

![image](https://user-images.githubusercontent.com/6471716/100161360-1a59c500-2ed3-11eb-91c2-908ddd61c5ba.png)

![image](https://user-images.githubusercontent.com/6471716/100161386-26de1d80-2ed3-11eb-80dd-a6e33628a2fa.png)

[SpriteEditor.zip](https://github.com/TheBoneJarmer/Arqan/files/5593271/SpriteEditor.zip)

**Comment by TheBoneJarmer** [comment]
Hi @FireBanana 

I have taken a look at your code but I have a hard time figuring out how you are trying to get a result. I have never used the glDrawElements function, but from what I can remember it used a part of a buffer using the index provided to tell the shader which vertices to use. Yet your code seems to do something different. In either case I recommend looking up tutorials on glDrawElements because I don't have enough experience with this.

That being said, could you elaborate what you are trying to do? Are you simply trying to make a quad render on the screen with a texture? Because if it helps, I recommend peeking in the source of Arqanore's Sprite class, which can be found [here](https://github.com/TheBoneJarmer/Arqanore/blob/master/src/Arqanore/Graphics/Sprite.cs). I don't use the glDrawElements function but I do use the glDrawArrays function to render a static or animated sprite on a quad.

**Comment by TheBoneJarmer** [comment]
That being said, I probably don't do things the correct way either, but they work so that is good enough for me. D:

**Comment by FireBanana** [comment]
Hey @TheBoneJarmer

glDrawElements can be used to render multiple triangles in one draw call as far as I know, saving on performance. I guess it's not necessary in this example because only one quad is being rendered but I was testing it out for future use. I have cross-checked it with similar C++ code so I'm unsure myself of why it doesn't display anything.

**Comment by TheBoneJarmer** [comment]
Hi @FireBanana 

Now you mention it, that does ring a bell. Thanks for explaining. That being said, could you share the c++ code that worked for you? I'll look at the glDrawElements function again to be 100% certain I did the signature correctly.

**Comment by FireBanana** [comment]
This is where it's replicated from:

https://learnopengl.com/code_viewer_gh.php?code=src/1.getting_started/4.1.textures/textures.cpp

**Comment by TheBoneJarmer** [comment]
Thanks! In the meantime I noticed something after looking at the docs from Khronos and the c++ example you shared. Your third argument, for the _type_ parameter, is incorrect. According to the [docs](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glDrawElements.xhtml) this should be GL_UNSIGNED_BYTE, GL_UNSIGNED_SHORT, or GL_UNSIGNED_INT. You are using a GL_FLOAT. Your current code is:
```c#
GL11.glDrawElements(GL11.GL_TRIANGLES, 0, GL11.GL_FLOAT, IntPtr.Zero);
```
But I believe it should be:

```c#
GL11.glDrawElements(GL11.GL_TRIANGLES, 0, GL11.GL_UNSIGNED_INT, IntPtr.Zero);
```
The c++ example you shared does it that way. Could you give this a shot and see if that works?

**Comment by FireBanana** [comment]
That's how I originally tried it, but didnt work. In the glBufferData in the C++ code, they pass in a uint array, but it was only possible to pass in a float array in C#, so I was testing it out with GL_FLOAT instead.

**Comment by TheBoneJarmer** [comment]
I have fiddled around a bit and looked a bit more thoroughly to the c++ example and I too cannot figure out in the world what is wrong. I tripple-checked the signatures of both methods, even introduced a new delegate for the glBufferData callback to use uints instead of floats, no success.

Did you manage to get the c++ example to run by the way? And did it work?

**Comment by TheBoneJarmer** [comment]
The code change was pushed to develop by the way. Will release a patch tomorrow.

**Comment by FireBanana** [comment]
Yeah the C++ one was running fine. I'll try to implement glGetError today and see if that returns anything

**Comment by TheBoneJarmer** [comment]
Ok, I just merged my changes to master so a patch should be available within a couple of minutes. I will take a look again today as well. This is proving to be a bigger challenge than I thought. But since the c++ example works, it have to be a signature mistake somewhere as the whole idea behind Arqan is that it looks a lot like it's C counterpart. And your C# code is almost a 100% copy.

**Comment by FireBanana** [comment]
Yeah, I believe that since its implemented similarly to glDrawArrays, there shouldn't be an error in the implementation. I rendered the quad right now using glDrawArrays and other than a slight offset it seems to be fine so the data/buffer implementation seems to be okay as well. 

![image](https://user-images.githubusercontent.com/6471716/100207851-e1007400-2f29-11eb-8056-3124be116237.png)

I'll still keep looking and let you know if I find anything. Also, glGetError() did not return any errors other than the GL_FLOAT that I used (which I have changed back)

**Comment by TheBoneJarmer** [comment]
Thanks for the heads up. I just had something weird. Yesterday when I tested my changes with Arqan, your code ran fine. Now my application crashes when I try to use glDrawArrays with Arqan v1.0.4. Can you give it a shot as well? If it stops working for you too, than I'll reverse my change. Although I would find it odd since you can pass both floats and uints to the glBufferData function.

**Comment by FireBanana** [comment]
Yep, throwing a null exception

![image](https://user-images.githubusercontent.com/6471716/100209003-2b362500-2f2b-11eb-8fe8-145afba23714.png)

**Comment by TheBoneJarmer** [comment]
Yea, I already figured it out in the meantime. Did not pay close attention to the delegate and screwed it up as result. But I guess those things can happen when it is 3AM in the morning and I still did not went to bed lol

In either case my apologies. Fixed it and I also added more overrides to the glBufferData function. Not just for floats or uints but for ints and doubles too. I thought, now I'm at it I might as well do it right. NuGet package should be available soon.

**Comment by FireBanana** [comment]
Haha no worries, thanks a lot for looking into it.

**Comment by TheBoneJarmer** [comment]
Oh I got some good news by the way. I just re-ran your code and it worked this time. :D It still renders one triangle in my case, which I find odd. Could you give it a shot with Arqan 1.0.5? It should be available by now or either it will soon.

**Comment by FireBanana** [comment]
Just tried it, working great. The whole quad renders for me.

**Comment by TheBoneJarmer** [comment]
Awesome! Glad to hear it works. In that case this was indeed because of a wrong signature in the delegate for the glBufferData function. Do you mind sharing the last source again by the way? I still like to see if I get a quad now too.

**Comment by FireBanana** [comment]
Here you go:

[SpriteEditor.zip](https://github.com/TheBoneJarmer/Arqan/files/5595982/SpriteEditor.zip)

The offset is still there strangely. Other than that, its perfect!

**Comment by TheBoneJarmer** [comment]
Awesome, now I get to see a quad as well. And I also see that offset issue what you experience. The texcoords look fine so that can't be it. The mipmap perhaps?

Anyways, there is something I need to tell. I am busy merging all methods and enums into a single static class, GL. Back when I generated these classes (somewhere in 2017), the OpenGL spec made a division between versions. But the latest OpenGL xml does not only contain GL46 classes but all commands are stored in one GL "namespace". Therefore I'd figured it'd be better if I did the same. So don't get a heart attack when you get a dozens of obsolete warnings. The current classes will not removed in the next release but they will be in the release after.

That being said, if there is anything else just let me know. I am happy that your code works now. :)
